Multistage economizer control

lymereJ opened the pull request:
Pull request overview
---------------------
 - Address #6109
 - This pull request adds a new feature to EnergyPlus that enables users to simulate an "economizer first" approach when modeling an economizer in a sensible-only load-controlled unitary systems with multispeed (or variable speed) fans and multispeed (or variable speed) coils. The new approach lets the economizer operation dictates the system flow rate instead of being decided based on the cooling coil speed (current approach). This new feature allows to simulate the performance of commercially available economizer controller for such systems more closely.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

lymereJ committed:
Add intial version of the NFP. [decent_ci_skip]

lymereJ committed:
Merge branch 'develop' into multistage_econ

lymereJ committed:
Add design document and small edits to the NFP.

lymereJ committed:
Merge branch 'develop' into multistage_econ

lymereJ committed:
Update NFP based on comments.

lymereJ committed:
Preliminary code changes.

lymereJ committed:
Clean up.

lymereJ committed:
Update mixed air temperature setpoint calcs.

lymereJ committed:
Adjust mixed air flow rate for autosized rated flow rate for VS coils.

lymereJ committed:
Add example file and warning when new feature is not being applied.

lymereJ committed:
Merge branch 'develop' into multistage_econ

lymereJ committed:
Do not perform verification if air loop index is not valid.

lymereJ committed:
Clang!

nrel-bot-2b commented:
@lymereJ @myoldmopar it has been 28 days since this pull request was last updated.

lymereJ committed:
Update MS economizer control: add logic when 100% OA at one speed would overcool and 100% at the lower speed would not be enough. The approach used is to run at both speeds to meet the load and average the flow rate. Add a new function to get the a fan DT at a specific mass flow rate and airflow ratio.

amirroth commented on src/EnergyPlus/DataHVACGlobals.hh:
```diff
@@ -150,6 +150,9 @@ namespace DataHVACGlobals {
     // OA Controller Heat Recovery Bypass Control Types
     int constexpr BypassWhenWithinEconomizerLimits(0);   // heat recovery controlled by economizer limits
     int constexpr BypassWhenOAFlowGreaterThanMinimum(1); // heat recovery ON at minimum OA in economizer mode
+    // OA Controller Economizer Staging
```
Make these into an enum and use `getEnumerationValue` (the pattern is in 200+ places already).

amirroth commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -791,6 +791,40 @@ void SimOAController(EnergyPlusData &state, std::string const &CtrlName, int &Ct
+                for (int BranchNum = 1; BranchNum <= primaryAirSystems.NumBranches; ++BranchNum) {
+                    for (int CompNum = 1; CompNum <= primaryAirSystems.Branch(BranchNum).TotalComponents; ++CompNum) {
+                        if (primaryAirSystems.Branch(BranchNum).Comp(CompNum).CompType_Num == SimAirServingZones::CompType::UnitarySystemModel) {
+                            std::string unitarySystemName = primaryAirSystems.Branch(BranchNum).Comp(CompNum).Name;
```
No need to make a copy of the string, use `const &` (or `std::string_view`).

amirroth commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -791,6 +791,40 @@ void SimOAController(EnergyPlusData &state, std::string const &CtrlName, int &Ct
+                            int unitarySystemNum = UtilityRoutines::FindItemInList(
+                                unitarySystemName, state.dataUnitarySystems->unitarySys, state.dataUnitarySystems->numUnitarySystems);
+                            if (state.dataUnitarySystems->unitarySys[unitarySystemNum - 1].m_ControlType ==
+                                UnitarySystems::UnitarySys::UnitarySysCtrlType::Load) {
```
Not part of this PR, but I would like to point out that in this fully-resolved name, `UnitarySys` appears THREE times! So, what you're telling me is that this is related to unitary systems?

amirroth commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -10646,35 +10659,44 @@ namespace UnitarySystems {
-                //      IF (UnitarySystem(UnitarySysNum)%AirFlowControl .EQ. UseCompressorOnFlow) THEN
                 if (this->m_FanOpMode == DataHVACGlobals::ContFanCycCoil) {
                     if (CoolSpeedNum == 0) {
                         state.dataUnitarySystems->CompOffMassFlow = this->MaxNoCoolHeatAirMassFlow;
```
Random questions, why are these essentially global variables?

lymereJ committed:
Unit tests, address comments, reformatting.

lymereJ committed:
Add unit tests.

lymereJ committed:
Merge develop.

lymereJ committed:
Revert test file changes.

lymereJ committed:
Revert test file changes.

lymereJ committed:
Update example file and minor fixes.

lymereJ committed:
Remove unused vars.

lymereJ committed:
Merge branch 'develop' into multistage_econ

lymereJ committed:
Address test failures.

lymereJ committed:
Handle confusing warnings triggered by multistage economizer operation.

lymereJ committed:
Add docs. [decent_ci_skip] [actions skip]

lymereJ commented on doc/input-output-reference/src/overview/group-controllers.tex:
```diff
@@ -295,6 +295,10 @@ \subsubsection{Inputs}\label{inputs-1-007}
 
+\paragraph{Field: Economizer Staging Operation}\label{field-economizer-staging-operation}
+
+This input is only applicable when the controller is used on an air loop that includes an \hyperref[airloophvacunitarysystem]{AirLoopHVAC:UnitarySystem} with multiple cooling speeds. When modeling an \hyperref[airloophvacunitarysystem]{AirLoopHVAC:UnitarySystem} with multiple cooling speeds, (as specified in a [unitarysystemperformancemultispeed]{UnitarySystemPerformance:Multispeed}), \textbf{Eco ...
```
During internal discussions the difference between continuously operating and cycling fan operations was brought up. After much thoughts, I'm proposing that we keep this approach for now. This type of equipment is typically controlled based on a primary and secondary thermostat inputs (see "Y1" and "Y2" in the manufacturer published sequences of operation included in the NFP), the higher fan speed being triggered when both inputs are "on". Since EnergyPlus doesn't really work that way, the proposed approach in this PR relies on the actual zone load. A follow-on feature could perhaps include a slightly modified approach for cycling fan operation only which would not always make the system run at the lowest available speed that would meet the load assuming economizer-only operation but aim for a higher speed based on how large the predicted load is (compared to a reference value). So, during a particular timestep the economizer could run at that higher speed only for part of it and cycle off for the rest of it. I'm not sure if this would significantly change the simulations results.

lymereJ commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -12730,8 +12730,9 @@ void CalcMultiSpeedDXCoilCooling(EnergyPlusData &state,
-                (VolFlowperRatedTotCap > state.dataHVACGlobal->MaxCoolVolFlowPerRatedTotCap(DXCT))) {
+            if (((VolFlowperRatedTotCap < state.dataHVACGlobal->MinOperVolFlowPerRatedTotCap(DXCT)) ||
+                 (VolFlowperRatedTotCap > state.dataHVACGlobal->MaxCoolVolFlowPerRatedTotCap(DXCT))) &&
+                state.dataHVACGlobal->MSUSEconoSpeedNum == 0) {
```
Only check the m3/s/W boundaries when the economizer is not running. Testing showed that without this, a rather large number of warnings would get published when for example the system flow rate corresponds to the high cooling speed flow rate and the cooling coil operates at its lowest speed. IMO, this could have confused users since the warning is written from a "point-of-view" of the cooling coil.

rraustad replied:
Ok, I see this comment. If the economizer is running then the coil air flow is on the high end of cfm/ton. As I said before this is not an issue in the real world. It may be time to remove those cfm/ton warnings and replace them with a minimum supply air temperature input field for the cooling coil models. That would be far more useful. I recently saw a -something C (-20C?) off a cooling coil in dehumidification mode. Yup, that would certainly dehumidify a zone when freezing air borne moisture onto the cooling coil.

lymereJ commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -794,6 +794,48 @@ void SimOAController(EnergyPlusData &state, std::string const &CtrlName, int &Ct
+            primaryAirSystems.EconomizerStagingCheckFlag = true;
+        }
+    }
+
```
Checking that the users are using the new input in the right context.

lymereJ commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -1132,6 +1132,14 @@ namespace UnitarySystems {
+                if (this->OAControllerIndex > 0) {
+                    this->OAControllerEconomizerStagingType = state.dataMixedAir->OAController(this->OAControllerIndex).EconomizerStagingType;
+                }
+            }
```
"Links" the controller outdoor air object to the unitary system.

lymereJ commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -8478,6 +8485,11 @@ namespace UnitarySystems {
+        if (this->OAControllerEconomizerStagingType == DataHVACGlobals::EconomizerStagingType::EconomizerFirst) {
+            manageEconomizerStagingOperation(state, AirLoopNum, FirstHVACIteration, ZoneLoad);
+        }
+
```
Point of origin for the economizer staging operation.

lymereJ commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -8550,103 +8562,84 @@ namespace UnitarySystems {
+        PartLoadRatio = 1.0;
+        this->FanPartLoadRatio = 1.0;
+        // Get full load result for non-multistage economizer operations
+        if (this->m_EconoSpeedNum == 0) {
```
Calculating the full load results when the economizer is running was generating some confusing errors for users.

For instance, when the economizer speed is low, these calculations are carried out at the highest cooling speed which result in negative enthalpy/temperature values at the outlet of the coil due to an unrealistic mix of large capacity and low air flow. In this case, warning would be issued about the negative values, however, the simulation would proceed (since the criteria for the `return` would not be met) and the actual PLR is calculated after. So, these warnings would have no real impact on the simulation and might just be a cause for concerns from users.

The proposed approach is to bypass the full load result calculations when MS economizer is active and operating, and let the code below determine the coil PLR (may it be equal to 1).

lymereJ commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -8660,53 +8653,77 @@ namespace UnitarySystems {
             }
+        } else {
+            // define the sensible load to meet for operation with multi-stage economizer
+            this->m_CoolingCoilSensDemand = std::abs(ZoneLoad);
```
The sensible load still needs to be defined when the MS economizer are active and operating, otherwise the coil won't turn on if necessary.

lymereJ commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -10626,35 +10643,44 @@ namespace UnitarySystems {
                     } else if (CoolSpeedNum == 1) {
                         state.dataUnitarySystems->CompOffMassFlow = this->m_CoolMassFlowRate[CoolSpeedNum];
-                        state.dataUnitarySystems->CompOffFlowRatio = this->m_CoolMassFlowRate[CoolSpeedNum];
+                        state.dataUnitarySystems->CompOffFlowRatio = this->m_MSCoolingSpeedRatio[CoolSpeedNum];
```
I think this was an error. The variable is expecting a flow ratio, not a flow rate.

lymereJ commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -10626,35 +10643,44 @@ namespace UnitarySystems {
-                    state.dataUnitarySystems->CompOnMassFlow = this->m_CoolMassFlowRate[1];
-                    state.dataUnitarySystems->CompOnFlowRatio = this->m_MSCoolingSpeedRatio[1];
+                } else if (this->m_EconoSpeedNum > 0) { // multi-stage economizer operation; set system flow rate to economizer flow rate
+                    state.dataUnitarySystems->CompOnMassFlow = this->m_CoolMassFlowRate[this->m_EconoSpeedNum];
+                    state.dataUnitarySystems->CompOnFlowRatio = this->m_MSCoolingSpeedRatio[this->m_EconoSpeedNum];
                     state.dataUnitarySystems->OACompOnMassFlow = this->m_CoolOutAirMassFlow;
-                } else if (CoolSpeedNum > 1) {
+                } else if (CoolSpeedNum > 0) {
```
Give priority to the economizer to choose the system flow rates. Note: `CompOn[...]` is used even when the economizer alone can meet the load (cooling coil is not actually being turned on).

lymereJ commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -10945,7 +10972,8 @@ namespace UnitarySystems {
         Real64 FanPartLoadRatio = PartLoadRatio;
         if (this->m_SimASHRAEModel) FanPartLoadRatio = this->FanPartLoadRatio;
-        int SpeedNum = max(this->m_CoolingSpeedNum, this->m_HeatingSpeedNum);
+        if (this->m_EconoPartLoadRatio > 0) FanPartLoadRatio = this->m_EconoPartLoadRatio;
+        int SpeedNum = max(this->m_CoolingSpeedNum, this->m_HeatingSpeedNum, this->m_EconoSpeedNum);
```
When `EconomizerFirst` is used, the `FanPartLoadRatio` is not based on the compressor part load ratio but based on the "economizer part load ratio" since the economizer operation dictates the system flow rate.

lymereJ commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -18058,5 +18092,147 @@ namespace UnitarySystems {
             });
     }
 
+    Real64 UnitarySys::getFanDeltaTemp(EnergyPlusData &state, bool const firstHVACIteration, Real64 const massFlowRate, Real64 const airFlowRatio)
+    {
+        Real64 fanDT = 0;
+        int FanInletNode = 0;
+        int FanOutletNode = 0;
+        if (this->m_FanType_Num == DataHVACGlobals::FanType_SystemModelObject) {
+            FanInletNode = state.dataHVACFan->fanObjs[this->m_FanIndex]->inletNodeNum;
+            FanOutletNode = state.dataHVACFan->fanObjs[this->m_FanIndex]->outletNodeNum;
+            state.dataLoopNodes->Node(FanInletNode).MassFlowRate = massFlowRate;
+            state.dataHVACFan->fanObjs[this->m_FanIndex]->simulate(state, airFlowRatio, _, _, _, _, _);
+        } else {
+            FanInletNode = state.dataFans->Fan(this->m_FanIndex).InletNodeNum;
+            FanOutletNode = state.dataFans->Fan(this->m_FanIndex).OutletNodeNum;
+            state.dataLoopNodes->Node(FanInletNode).MassFlowRate = massFlowRate;
+            Fans::SimulateFanComponents(state, blankString, firstHVACIteration, this->m_FanIndex, airFlowRatio);
+        }
+        fanDT = state.dataLoopNodes->Node(FanOutletNode).Temp - state.dataLoopNodes->Node(FanInletNode).Temp;
+        return fanDT;
+    }
```
New function to calculate a fan's node temperature differential based on a specific mass flow and air flow ratio.

lymereJ commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -18058,5 +18092,147 @@ namespace UnitarySystems {
+        return fanDT;
+    }
+
+    void UnitarySys::setEconomizerStagingOperationSpeed(EnergyPlusData &state, bool const firstHVACIteration, Real64 const zoneLoad)
```
New function that determines at which speed the economizer should run to meet the current load.

lymereJ commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -18058,5 +18092,147 @@ namespace UnitarySystems {
+        }
+    }
+
+    void UnitarySys::calcMixedTempAirSPforEconomizerStagingOperation(EnergyPlusData &state,
```
New function that requests an updated mixed air flow rate based on the predetermined economizer speed and recalculates the mixed air setpoint to meet the load.

lymereJ commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -18058,5 +18092,147 @@ namespace UnitarySystems {
+    }
+
+    void
+    UnitarySys::manageEconomizerStagingOperation(EnergyPlusData &state, int const airLoopNum, bool const firstHVACIteration, Real64 const zoneLoad)
```
New functions that orchestrates the economizer staging operation: 1) determine economizer speed, 2) recalculate mixed air conditions and setpoint, 3) recalculate outdoor air fraction.

lymereJ commented on testfiles/CMakeLists.txt:
```diff
@@ -735,6 +735,7 @@ add_simulation_test(IDF_FILE WCE_DoubleClear_BSDF.idf EPW_FILE USA_IL_Chicago-OH
 add_simulation_test(IDF_FILE WCE_Interior_VB_-45_deg.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE _5ZoneAirCooled_annual.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw ANNUAL_SIMULATION)
 add_simulation_test(IDF_FILE _5ZoneAirCooled_LeapYear_annual.idf EPW_FILE MadeUpLeapYear.epw ANNUAL_SIMULATION)
+add_simulation_test(IDF_FILE UnitarySystem_MultiSpeedDX_EconoStaging.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw ANNUAL_SIMULATION)
```
New example file which uses the new input in 4 different configurations:
- `Coil:Cooling:DX` and `Fan:VariableVolume`
- `Coil:Cooling:DX` and `Fan:SystemModel` (cycling)
- `Coil:Cooling:DX:MultiSpeed` and `Fan:VariableVolume`
- `Coil:Cooling:DX:VariableSpeed` and `Fan:VariableVolume`

lymereJ commented on tst/EnergyPlus/unit/UnitarySystem.unit.cc:
```diff
@@ -843,6 +844,11 @@ TEST_F(ZoneUnitarySysTest, Test_UnitarySystemModel_factory)
     EXPECT_NEAR(totalAncillaryPower, thisSys->m_TotalAuxElecPower, 0.00000001);
     // at PLR very near 0.5, m_TotalAuxElecPower should be very near 75 W.
     EXPECT_NEAR(73.96002, thisSys->m_TotalAuxElecPower, 0.0001);
+    // delta temperature calculations
+    Real64 fanDT = thisSys->getFanDeltaTemp(*state, true, 1, 1.0);
+    EXPECT_NEAR(fanDT, 0.7070, 0.0001);
+    fanDT = thisSys->getFanDeltaTemp(*state, true, 0.5, 0.5);
+    EXPECT_NEAR(fanDT, 0.7070, 0.0001);
```
Check for the new `getFanDeltaTemp` function. Here we have a constant volume fan the we expect the output to be the same regardless of the mass flow rate and air flow fraction we throw at it.

lymereJ commented on tst/EnergyPlus/unit/UnitarySystem.unit.cc:
```diff
@@ -13229,6 +13249,12 @@ TEST_F(EnergyPlusFixture, UnitarySystemModel_SizingWithFans)
     thisSys.m_MaxNoCoolHeatAirVolFlow = DataSizing::AutoSize;
     thisSys.m_DesignFanVolFlowRate = DataSizing::AutoSize;
 
+    // delta temperature calculations
+    Real64 fanDT = thisSys.getFanDeltaTemp(*state, true, 1, 1.0);
+    EXPECT_NEAR(fanDT, 0.3068, 0.0001);
+    fanDT = thisSys.getFanDeltaTemp(*state, true, 0.5, 0.5);
+    EXPECT_NEAR(fanDT, 0.17615, 0.0001);
```
Check for the new `getFanDeltaTemp` function. Here we have a multispeed fan so we expect the output to be different at a lower mass flow rate and air flow ratio.

lymereJ commented on tst/EnergyPlus/unit/UnitarySystem.unit.cc:
```diff
@@ -20911,6 +20937,116 @@ Coil:Heating:Gas:MultiStage,
     EXPECT_FALSE(state->dataUnitarySystems->CoolingLoad);
 }
 
+TEST_F(EnergyPlusFixture, SetEconomizerStagingOperationSpeedTest)
```
New test for the new functions that were introduced.

lymereJ committed:
Fix typo. [decent_ci_skip] [actions skip]

lymereJ commented:
All big diffs are due to the fact that a new default value was added, which is expected. This also shows up on some of the small diffs (table). There are a limited number of ERR, EIO, MTR, SSZ, and ZSZ. The numerical differences for the few MTR, SSZ, and ZSZ are very small. The EIO diffs are due to slightly different warmup convergence results. The ERR results are due to a slight change in the number of occurrence of pre-existing warnings.

lymereJ commented:
The figure below shows two of days simulated with the current develop and this branch. The bottom panel shows the space/zone temperature for both cases. The expected behavior is produced: this system has two cooling speeds, in the proposed case the economizer ramps up to meet the load this decreases the need for mechanical cooling.

![image](https://github.com/NREL/EnergyPlus/assets/5149279/0fa7a2a4-abc5-4a21-a74b-ed28a1626d52)

Here is the end-use breakdown comparison for the new example file ran in Seattle, WA (4C). As expected we see a trade-off between mechanical cooling and fan energy use, that being said, the proposed approach (which uses the new `EconomizerFirst` input) shows less energy use overall for a similar zone temperature control.
- Current
<img src="https://github-production-user-asset-6210df.s3.amazonaws.com/5149279/247215074-a94ea736-d05a-4871-a46d-2b5fb97487ca.png" width="280" height="450">

- Proposed
<img src="https://github.com/NREL/EnergyPlus/assets/5149279/01f58f68-8f23-4921-96b3-02d9b0b4ac64" width="280" height="450">

nrel-bot-3 commented:
@lymereJ @myoldmopar it has been 28 days since this pull request was last updated.

nrel-bot-2b commented:
@lymereJ it has been 9 days since this pull request was last updated.

nrel-bot-2 commented:
@lymereJ it has been 7 days since this pull request was last updated.

Myoldmopar commented:
Summarizing the status here:
 - Last time this ran, CI had one debug integration failure, but it was unrelated.  A fresh build would presumably be clean on the debug runs
 - There are a good number of diffs in this branch, including small math and EIO diffs.  I don't see an explanation/analysis/review of these, but maybe that was done within a commit and not showing up on this page.  That would be most welcome.  Perhaps just a fresh run with updated develop pulled in would clean some?  Not sure.
 - The branch is conflicted in UnitarySystem.  I'm not surprised, given how many coil related changes we've merged in the last week.  I don't intend on resolving them, but can assist if needed.  
 - Given how this changes/enhances control in Unitary, I think it would be vital to get @rraustad to bless this before we move toward a merge queue.

I'm going to take a quick attempt at pulling develop in and resolving the conflicts.  If it's not crazy, I'll get them done and push up to get fresh CI results.

lymereJ commented:
> * There are a good number of diffs in this branch, including small math and EIO diffs.  I don't see an explanation/analysis/review of these, but maybe that was done within a commit and not showing up on this page.  That would be most welcome.  Perhaps just a fresh run with updated develop pulled in would clean some?  Not sure.

I had provided this summary below for the diffs. I can definitely expand on it if need be once develop has been merged in and the conflicts have been resolved.

> All big diffs are due to the fact that a new default value was added, which is expected. This also shows up on some of the small diffs (table). There are a limited number of ERR, EIO, MTR, SSZ, and ZSZ. The numerical differences for the few MTR, SSZ, and ZSZ are very small. The EIO diffs are due to slightly different warmup convergence results. The ERR results are due to a slight change in the number of occurrence of pre-existing warnings.

I'm happy to help with resolving the conflicts.

Myoldmopar commented:
I see the comment about diffs now, thanks.  For the existing files, what change is it that is causing the small numeric change?  It's always nice to be able to "prove" no-diffs in cases where we shouldn't be affecting the results.  But sometimes it's just simply not possible, and that's OK.

Conflicts: Yeah, they are not huge -- but they do seem to be interconnected with logic in the other recent changes.  I'm not privy enough on the  individual changes to confidently integrate them together without doing a deep dive.  Which I don't have time for right now.  If you (@lymereJ) can you take a look at the conflicts when pulling in latest develop, it may be obvious to you.  If it's not, we may need to gather some folks and do it live together.

lymereJ commented:
> I see the comment about diffs now, thanks. For the existing files, what change is it that is causing the small numeric change? It's always nice to be able to "prove" no-diffs in cases where we shouldn't be affecting the results. But sometimes it's just simply not possible, and that's OK.
 
When you say small numeric changes are you referring to the diffs in files other than the tables? For the tables, I believe that most of the small diffs are due to the fact that we added a new input with defaults.

![image](https://github.com/NREL/EnergyPlus/assets/5149279/87a75780-e8e2-4d60-96de-3e536acde98c)

> Conflicts: Yeah, they are not huge -- but they do seem to be interconnected with logic in the other recent changes. I'm not privy enough on the individual changes to confidently integrate them together without doing a deep dive. Which I don't have time for right now. If you (@lymereJ) can you take a look at the conflicts when pulling in latest develop, it may be obvious to you. If it's not, we may need to gather some folks and do it live together.

Sounds good! I'll take a shot at it later today. If we do need a "live session" to address the conflicts, could these changes still make it into the next release?

Myoldmopar commented:
I'm talking about the EIO and ESO diffs which were present on the last run that Decent did on this branch (https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f3899cf9907230c2f7407bf7c4e308cf17b831df-x86_64-MacOS-10.17-clang-13.0.0.html).

Once you take a pass over the conflicts, let me know this afternoon how it went.  We can try to coordinate something tomorrow if we need to do a little 15 minute group session to resolve conflicts live together.  As long as we keep momentum on it, I think we should be able to get it in this release.

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -9514,7 +9524,7 @@ namespace UnitarySystems {
                                 }
                                 TempMinPLR = TempMaxPLR;
-                                while ((TempSysOutput - TempLoad) < 0.0 && TempMinPLR > 0.05) {
+                                while ((TempSysOutput - TempLoad) < 0.0 && TempMinPLR > 0.025) {
```
@lymereJ why was this changed? The PLR increment is 0.1, 0.05 splits the difference between 0.0 and 0.1. Did changing this to 0.025 correct something? This doesn't really bother me because 0.025 is still between 0.0 and 0.1. It's more of a curiosity that I ask. Also, this is way into the weeds in UnitarySystem so I have to think you stepped through the debugger for some time looking for issues in order to get here. I am guessing you don't recall why you changed this but am hoping you do recall.

lymereJ replied:
Yes, It's been a while but if I recall correctly, I observed that when running the economizer at high speed, sometimes not all the load could be met and that the cooling coil had to run and in some of these instances the load left to be met by the coil was very small and so would be the PLR. Because of that limit at 0.05 I was getting some convergence warning, I believe that lowering the min PLR here helped with that.

rraustad replied:
Oh yeah, I didn't look far enough down the code. The increment on minPLR is 0.01 (I recall that now, 0.1 on max, then reduce to 0.01 on min). So this could (should?) be lowered to 0.005 (I might have mistakenly left a 0 out when I wrote this years ago). Since you have tested this I guess leave if for now and move on. If you have other changes to make please put a comment here that the lower limit might be changed to 0.005 without adverse affects.

                                while ((TempSysOutput - TempLoad) < 0.0 && TempMinPLR > 0.025) {
                                    // pull upper limit of HeatingPLR down to last valid limit (i.e. heat output still exceeds SystemSensibleLoad)
                                    TempMaxPLR = TempMinPLR;
                                    // find minimum limit of HeatingPLR
                                    TempMinPLR -= 0.01;

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -10747,35 +10757,44 @@ namespace UnitarySystems {
+                    state.dataUnitarySystems->CompOnFlowRatio = this->m_MSCoolingSpeedRatio[this->m_EconoSpeedNum];
                     state.dataUnitarySystems->OACompOnMassFlow = this->m_CoolOutAirMassFlow;
-                } else if (CoolSpeedNum > 1) {
+                } else if (CoolSpeedNum > 0) {
```
This logical stack is interesting. I thought the intent was to allow economizer operation without compressor operation. Here, if the compressor is off the code won't make it to economizer operation (unless I am not interpreting `CoolSpeedNum == 0` correctly, which as you now know, is easy to do).

                if (CoolSpeedNum == 0) {
                } else if (this->m_EconoSpeedNum > 0) { // multi-stage economizer operation
                } else if (CoolSpeedNum > 0) {

I recall the plots and thought it looked like everything was working correctly, but looking at this I wonder if that's true for both choices. It looks like this would work for InterlockedWithMechanicalCooling, but for EconomizerFirst how could the economizer ramp up when the compressor was off? It looks like the first 2 stacks should be switched (with added conditionals, or maybe an EconomizerFirst to start this stack off with), but remember, I'm looking at this on git and not in the debugger.

    A20; \field Economizer Operation Staging
       \type choice
       \key EconomizerFirst
       \key InterlockedWithMechanicalCooling

lymereJ replied:
Yes, this is confusing but unless I am missing something `m_CoolingSpeedNum` (and hence `CoolSpeedNum`) is at least `> 1` if `m_EconoSpeedNum` is `> 1`, see the snippet below:

```
        this->m_EconoSpeedNum = 0;
        if (this->OAControllerEconomizerStagingType == DataHVACGlobals::EconomizerStagingType::EconomizerFirst) {
            manageEconomizerStagingOperation(state, AirLoopNum, FirstHVACIteration, ZoneLoad);
        }

        // if a variable speed unit, the SensOutputOff at SpeedNum=1 must be checked to see if it exceeds the ZoneLoad
        // This is still no load but at the first speed above idle
        if ((state.dataUnitarySystems->HeatingLoad && this->m_NumOfSpeedHeating > 0) ||
            (state.dataUnitarySystems->CoolingLoad && this->m_NumOfSpeedCooling > 0)) {
            if (this->m_Staged) {
                if (state.dataUnitarySystems->HeatingLoad) {
                    this->m_HeatingSpeedNum = this->m_StageNum;
                } else {
                    this->m_CoolingSpeedNum = std::abs(this->m_StageNum);
                }
            } else {
                if (state.dataUnitarySystems->HeatingLoad) {
                    this->m_HeatingSpeedNum = 1;
                } else {
                    this->m_CoolingSpeedNum = 1;
                }
            }
            // calcUnitarySystemToLoad calls setOnOffMassFlowRate so probably no need to call this here
            this->setOnOffMassFlowRate(state, OnOffAirFlowRatio, PartLoadRatio);
```

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -11066,7 +11086,8 @@ namespace UnitarySystems {
         if (this->m_SimASHRAEModel) FanPartLoadRatio = this->FanPartLoadRatio;
-        int SpeedNum = max(this->m_CoolingSpeedNum, this->m_HeatingSpeedNum);
+        if (this->m_EconoPartLoadRatio > 0) FanPartLoadRatio = this->m_EconoPartLoadRatio;
+        int SpeedNum = max(this->m_CoolingSpeedNum, this->m_HeatingSpeedNum, this->m_EconoSpeedNum);
```
So economizer fan speed gets top priority here. That makes sense. And I don't think SimASHRAEModel works with multispeed coils (I actually don't recall, it's been a while) so this seems OK. It's also good you understand what's happening here, the fan PLR and coil PLR are independent here. I wonder how this interacts with the DXCoil flow per ton limits warnings ?

lymereJ replied:
I did add some DXCoil flow per ton limit warning "bypass" when the `EconomizerFirst` option is used. I added a comment for that above: https://github.com/NREL/EnergyPlus/pull/9987#discussion_r1235521796.

rraustad replied:
That's fine. The flow per ton warning is not useful in the real world. There is an issue on very low flow per ton regarding coil freeze. A high flow per ton doesn't do anything other than waste fan power. And coil freeze can be tested using the coil outlet temp. So again, those warnings are mostly useless.

lymereJ commented:
> Once you take a pass over the conflicts, let me know this afternoon how it went. We can try to coordinate something tomorrow if we need to do a little 15 minute group session to resolve conflicts live together. As long as we keep momentum on it, I think we should be able to get it in this release.

Well, I broke everything, but I think that I know what I did wrong, I'll reset to the last good commit, re-address the conflicts, and test locally first before pushing.

rraustad commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -794,6 +794,48 @@ void SimOAController(EnergyPlusData &state, std::string const &CtrlName, int &Ct
+                        }
+                    }
+                }
+                if (sensLoadCtrlUnitarySystemFound == false) {
```
if (!sensLoadCtrlUnitarySystemFound)

lymereJ replied:
Thanks! Addressed in the latest commit.

rraustad commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -2364,6 +2406,17 @@ void ProcessOAControllerInputs(EnergyPlusData &state,
+            }
+        }
+    }
+
```
AlphArray(20) could be blank since it's not required. But if (NumAlphas > 19) then that field is present and the default is InterlockedWithMechanicalCooling

    A20; \field Economizer Operation Staging
       \type choice
       \key EconomizerFirst
       \key InterlockedWithMechanicalCooling
       \default InterlockedWithMechanicalCooling        

And this variable is initialized to interlock so this looks OK (I thought this was wrong until I looked at this).

    struct OAControllerProps

    DataHVACGlobals::EconomizerStagingType EconomizerStagingType =
            DataHVACGlobals::EconomizerStagingType::InterlockedWithMechanicalCooling; // User input select

rraustad commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -3096,6 +3149,7 @@ void InitOAController(EnergyPlusData &state, int const OAControllerNum, bool con
                 // if flow rate has been specified by a manager, set it to the specified value
                 thisOAController.MixMassFlow =
                     state.dataAirLoop->AirLoopFlow(AirLoopNum).ReqSupplyFrac * state.dataAirLoop->AirLoopFlow(AirLoopNum).DesSupply;
+                state.dataLoopNodes->Node(thisOAController.RetNode).MassFlowRate = thisOAController.MixMassFlow - thisOAController.ExhMassFlow;
```
No, no, no. @mjwitte you should look at this. We should not be changing the return node of the mixer.

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -11076,7 +11097,12 @@ namespace UnitarySystems {
+                    }
                     AverageUnitMassFlow = tempSpeedRatio * state.dataUnitarySystems->CompOnMassFlow +
                                           (1.0 - tempSpeedRatio) * state.dataUnitarySystems->CompOffMassFlow;
                 }
```
This one is hard to comprehend. This is an air flow calculation. And it either uses the econo PLR or the system fan PLR. Do you see how important the SetOnOffMassFlowRate function is? Everything has to be in sync. 

Since you do this right out of the gate in setAverageAirFlow then this looks OK.

        if (this->m_EconoPartLoadRatio > 0) FanPartLoadRatio = this->m_EconoPartLoadRatio;

Oh, wait a minute. FanPartLoadRatio is a local in setAverageAirFlow. So then shouldn't line 11102 be:

    tempSpeedRatio = this->m_EconoPartLoadRatio;

I searched for m_EconoPartLoadRatio and don't see where these (fanPLR and econoPLR) are equated in code.

lymereJ replied:
Just above we define the relationship between `m_EconoPartLoadRatio` and `FanPartLoadRatio` which I think is equivalent to what you propose.
```
if (this->m_EconoPartLoadRatio > 0) FanPartLoadRatio = this->m_EconoPartLoadRatio;
```

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -11221,8 +11247,9 @@ namespace UnitarySystems {
+            CoolingCompOn = DataHVACGlobals::CompressorOperation::Off;
+        } else if (this->m_CoolingSpeedNum > 1) { // for multispeed coils, comp is on IF speed > 1
             CoolingCompOn = DataHVACGlobals::CompressorOperation::On;
         }
```
I think that's trick how you inserted the economizer control into these IF blocks. It's just hard to know how the logic responds to both choices of econo control.

    if (CoolPLR > 0) {                 <- compressor is on
    } else if (CoolPLR == 0 && this->m_EconoSpeedNum > 0) { // turn compressor off for economizer calcs
    } else if (this->m_CoolingSpeedNum > 1) { // for multispeed coils, comp is on IF speed > 1

Like I said on the conference call, these new concepts are hard to nail down initially and it takes a few release cycles to figure out what we did wrong. And we don't know that until users come back and say "Hey, what the heck is going on here?"

From this inspection, this one looks OK (famous last words).

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -17856,5 +17883,147 @@ namespace UnitarySystems {
+            state, firstHVACIteration, mixedAirFlowRate, mixedAirFlowRate / this->m_CoolMassFlowRate[this->m_NumOfSpeedCooling]);
+        // determine new mixed air setpoint
+        Real64 newMixedAirSP = zoneTemp - std::abs(zoneLoad) / (cpAir * mixedAirFlowRate);
+        state.dataLoopNodes->Node(mixedAirNode).TempSetPoint = newMixedAirSP - fanDTAtMixedAirFlowRate;
```
This is scary and intriguing at the same time. UnitarySystem is literally setting the air loop air flow characteristics. LoopFlowRateSet, ReqSupplyFrac, and Node(mixedAirNode).TempSetPoint. I guess the air loop parent does that anyway so this might just work. This may be a good path forward to knowing what the OA system is doing inside UnitarySystem, because UnitarySystem is setting the stage. I won't even begin to question how this will work with AirflowNetwork. The problem I had when I tried this before was that the mixer return node condition is unknown (it's last iterations return condition), and that condition would not be known until the zones were simulated at this new air flow rate. And then the OA system better respond correctly. Things like this can easily result in oscillations, but that won't be known for a few release cycles. I am going to request that you show new plots with your example file, same as you did before, using both choices for economizer control (i.e., 2 plots).

lymereJ replied:
I'll provide the plots in a comment below.

rraustad commented:
@Myoldmopar I have reviewed, and provided some comments. There is some trick stuff going on here and I'll give those things a 10. I especially like how economizer control was inserted in the air flow IF blocks (whether or not they are correct at this point). I am a little hesitant on actually setting air system air flow characteristics and I guess OA controller properties (mixed air node SP), but hey, think outside the box. Another 10 in my opinion, even if it doesn't work. I'll wait to see some final plots of the econo choices before I approve. I hope those plots come through clean in about an hours time without seeing something of note. I also looked at diffs and don't see any real differences. Point me to a file that concerns you, if I missed something that is.

lymereJ commented:
Thanks for the review! I'll provide the plots later when I can get back to my desk.

On Thu, Aug 17, 2023, 6:20 PM Richard Raustad ***@***.***>
wrote:

> @Myoldmopar <https://github.com/Myoldmopar> I have reviewed, and provided
> some comments. There is some trick stuff going on here and I'll give those
> things a 10. I especially like how economizer control was inserted in the
> air flow IF blocks (whether or not they are correct at this point). I am a
> little hesitant on actually setting air system air flow characteristics and
> I guess OA controller properties (mixed air node SP), but hey, think
> outside the box. Another 10 in my opinion, even if it doesn't work. I'll
> wait to see some final plots of the econo choices before I approve. I hope
> those plots come through clean in about an hours time without seeing
> something of note. I also looked at diffs and don't see any real
> differences. Point me to a file that concerns you, if I missed something
> that is.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NREL/EnergyPlus/pull/9987#issuecomment-1683191764>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHJEX5TXFCE2GD4K67DLVLXV27PHANCNFSM6AAAAAAXPVCMUM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

lymereJ committed:
Correctly merge develop and address conflicts.

lymereJ committed:
Update example file following the input changes from develop.

lymereJ committed:
Address review comments.

lymereJ committed:
Clang!

lymereJ commented:
> Well, I broke everything, but I think that I know what I did wrong, I'll reset to the last good commit, re-address the conflicts, and test locally first before pushing.

Ok, I reverted the previous merge and re-did it correctly, _I think._ I also updated the new example file following the latest changes to develop, and made changes based on Richard's comments. The example file ran fine locally, I'm hoping that the CI will come back "clean" (as in the same types/magnitude of diffs as before).

@rraustad, here are the plots you requested for three consecutive days. Outdoor air is in orange for the `InterlockedWithMechanicalCooling` and maroon for the `EconomizerFirst`. Blue is the supply fan air flow rate for the `EconomizerFirst` case. The bottom panel shows the cooling rate.
- System 1

![image](https://github.com/NREL/EnergyPlus/assets/5149279/e09ff37d-1bde-4b92-a170-8e460b24024c)

- System 2

![image](https://github.com/NREL/EnergyPlus/assets/5149279/7795b9b1-bffb-4490-a943-a2f4606ba499)

- System 3

![image](https://github.com/NREL/EnergyPlus/assets/5149279/7a04c2a4-dd1d-4d19-854e-1f2bb3906299)

- System 4

![image](https://github.com/NREL/EnergyPlus/assets/5149279/a6a262b3-7826-4c84-b28e-3d0947941e08)

rraustad commented:
@lymereJ what I was looking for was an increase in fan flow without compressor operation. It appears to me that all these plots show that econoFirst does that. @Myoldmopar from what I see this branch is doing exactly what is intended. 

The only pause I have is this line in MixedAir. The mixer return node is set by the sum of the zone returns. I don't think that should be calculated/changed in the mixer code. It may be that this is the correct flow rate for this calculation of the mixer node flows, but I think the correct flow rate will wind up at the mixer return node naturally. It's the concept of stomping on what other components have already set.

    state.dataLoopNodes->Node(thisOAController.RetNode).MassFlowRate = thisOAController.MixMassFlow - thisOAController.ExhMassFlow;

@lymereJ if this line is deleted, what happens?

Myoldmopar commented:
Thanks @rraustad, that's great to hear!  I will kick off a local build of this branch and if we don't hear back from @lymereJ I can even setup a second build with that line removed and run tests/regressions.  

Other than that, CI results look much better, only very minor small diffs and warmup convergence style changes.  So I agree this is probably very close.

rraustad commented:
The mixer return node is the air loop supply side return, which is "connected" to the demand side outlet node. That node, the mixer return, is set/updated by the air loop manager and is a node that is tested for convergence. This is from the new example file UnitarySystem_MultiSpeedDX_EconoStaging.

    OutdoorAir:Mixer,
      Sys 1 Furnace DX Cool OA Mixing Box,  !- Name
      Sys 1 Furnace DX Cool Mixed Air Outlet,  !- Mixed Air Node Name
      Sys 1 Furnace DX Cool Outdoor Air Inlet,  !- Outdoor Air Stream Node Name
      Sys 1 Furnace DX Cool Relief Air Outlet,  !- Relief Air Stream Node Name
      Sys 1 Furnace DX Cool Air Loop Inlet;  !- Return Air Stream Node Name   <-- this node

    AirLoopHVAC,
      Sys 1 Furnace DX Cool,   !- Name
      ,                        !- Controller List Name
      Sys 1 Furnace DX Cool Availability Managers,  !- Availability Manager List Name
      autosize,                !- Design Supply Air Flow Rate {m3/s}
      Sys 1 Furnace DX Cool Branches,  !- Branch List Name
      ,                        !- Connector List Name
      Sys 1 Furnace DX Cool Air Loop Inlet,  !- Supply Side Inlet Node Name  <-- is this node
      Sys 1 Furnace DX Cool Return Air Outlet,  !- Demand Side Outlet Node Name
      Sys 1 Furnace DX Cool Supply Path Inlet,  !- Demand Side Inlet Node Names
      Sys 1 Furnace DX Cool Heating Coil Outlet;  !- Supply Side Outlet Node Names

That node gets updated/set in:

    HVACInterfaceManager::UpdateHVACInterface(state,
                                              ZoneGroupNum,
                                              DataConvergParams::CalledFrom::AirSystemDemandSide,
                                              state.dataAirLoop->AirToZoneNodeInfo(ZoneGroupNum).ZoneEquipReturnNodeNum(RetAirPathNum),
                                              state.dataAirLoop->AirToZoneNodeInfo(ZoneGroupNum).AirLoopReturnNodeNum(RetAirPathNum),
                                              SimAir);

Myoldmopar commented:
I just built this branch and ran it happily, all tests pass.  I did a second build of this branch where I commented that line out, once again it built and ran all tests passing.  I then ran a regression between the two branches and got zero diffs whatsoever.  I think you are right that that line is not necessary, so we can just comment/remove that line and maybe this will be ready to merge.  If that's correct, I'll push that up shortly, or @lymereJ feel free to chime in.

rraustad commented:
I'm surprised there were no diffs, but then that's a very good thing. Delete that line and then this branch can merge.

Myoldmopar committed:
Remove unnecessary flow rate override per @rraustad

Myoldmopar commented:
Pushed :+1:

Myoldmopar commented:
I'll let the basic GHA runs get going then merge shortly.

Myoldmopar commented:
I was hoping to catch it before Decent got started, but missed it.  I'm going to merge this but I"ll leave the branch temporarily so that Decent can finish its runs.  Thanks @lymereJ for this enhancement.  Thanks @rraustad for so much review; it seems like this PR is triggering some interesting refactoring ideas to me.

lymereJ commented:
Sorry for only getting back now. I wish I would have better documented the reason behind adding that line, right now, I got nothing. Perhaps it's a leftover from me testing things. I was going to suggest to comment it out to leave a trace in case an issue comes up but it looks like that's exactly what you did, thanks! For the record, I did run a full annual simulation for the new example file, the table files are the same, and I see some very negligible diffs in the ESO.

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -17856,5 +17891,147 @@ namespace UnitarySystems {
+        auto outdoorAirController = state.dataMixedAir->OAController(this->OAControllerIndex);
+        using Psychrometrics::PsyCpAirFnW;
+        Real64 cpAir = PsyCpAirFnW(state.dataLoopNodes->Node(outdoorAirController.InletNode).HumRat);
+        Real64 zoneTemp = state.dataLoopNodes->Node(state.dataZoneEquip->ZoneEquipConfig(this->ControlZoneNum).ZoneNode).Temp;
```
@lymereJ if the UnitarySystem is set up for Setpoint control, this->ControlZoneNum will be 0. Can you quickly test that to make sure there is no crash here, or check for Load control early on so this doesn't crash?

lymereJ replied:
@rraustad - We do check for load control early on but we just use the outcome to display a warning the system is controlled by setpoint, so I think it will crash. I can add a check for type of control in the if statement below so we don't perform economizer staging calculations if the system is setpoint controlled.
From:
```
this->m_EconoSpeedNum = 0;
if (this->OAControllerEconomizerStagingType == DataHVACGlobals::EconomizerStagingType::EconomizerFirst) {
    manageEconomizerStagingOperation(state, AirLoopNum, FirstHVACIteration, ZoneLoad);
}
```
To:
```
this->m_EconoSpeedNum = 0;
if (this->OAControllerEconomizerStagingType == DataHVACGlobals::EconomizerStagingType::EconomizerFirst &&
    this->m_ControlType == UnitarySysCtrlType::Load) {
    manageEconomizerStagingOperation(state, AirLoopNum, FirstHVACIteration, ZoneLoad);
}
```
Would that work for you?

lymereJ replied:
I checked and I don't think we need to do anything here. `manageEconomizerStagingOperation` is only called from `controlUnitarySystemOutput` which is only used for load based systems. So if a model uses the new `Controller:OutdoorAir` input, it won't change anything to the simulation. Moreover, the PR added a warning message to let the user know that that if that input is selected in that context it won't do anything.

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -17856,5 +17891,147 @@ namespace UnitarySystems {
+            state, firstHVACIteration, mixedAirFlowRate, mixedAirFlowRate / this->m_CoolMassFlowRate[this->m_NumOfSpeedCooling]);
+        // determine new mixed air setpoint
+        Real64 newMixedAirSP = zoneTemp - std::abs(zoneLoad) / (cpAir * mixedAirFlowRate);
+        state.dataLoopNodes->Node(mixedAirNode).TempSetPoint = newMixedAirSP - fanDTAtMixedAirFlowRate;
```
@lymereJ this new control puts a SP on the mixed air node that tries to control the OA system to meet the zone load. That's fine, but 1) is stomping on the mixed air SPM, and 2) the user does not know anything about this override. In fact I'm not even sure this is doing anything because the SPMs and OA system are simulated before the UnitarySystem. Here's another case where I want to know what happens if this SP control is removed. If it's necessary then fine, but the user needs to know what is happening in their simulation. If this in fact is needed then I think this should be described in the docs and the "Unitary System Mixed Air Setpoint Temperature" be reported out to the user.

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -17856,5 +17891,147 @@ namespace UnitarySystems {
+
+                    // recalculate the outdoor air fraction to meet the new mixed air setpoint at the new mixed air flow rate
+                    MixedAir::ManageOutsideAirSystem(
+                        state, state.dataAirLoop->OutsideAirSys(this->OASysIndex).Name, firstHVACIteration, airLoopNum, this->OASysIndex);
```
Oh, I see, the OA system gets called from the UnitarySystem. I previously thought that only the air loop control properties were set in UnitarySystem. So that means the mixed air set point IS getting used, and that calculation of that SP is necessary. This is pointing to a new report for "Unitary System Mixed Air Setpoint Temperature".

lymereJ replied:
I was just going through your comments one more time, so just to confirm, there's is nothing to add here, is that correct? The user should be able to get the setpoint set by the unitary system by requesting the system node setpoint temperature output variable.

rraustad replied:
I think viewing the SP on the node will be sufficient for now.

Myoldmopar commented:
It _does_ look like there is one fatal error, but only showing up on Windows oddly: https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-273aede840425898126ac7f34ce658436f17297f-Win64-Windows-10-VisualStudio-16.html#test

Please try to reproduce that and fix it when you push any additional commits.

lymereJ commented:
@Myoldmopar - Really odd, I ran the test on my Windows machine and it passes.

rraustad reviewed (commented):
Lot's of work here to include OA system operation with UnitarySystem performance. Nice effort.
